pkg/pool-utils/contracts/RecoveryMode.sol

```diff
@@ -85,13 +86,24 @@ abstract contract RecoveryMode is IRecoveryMode, BasePoolAuthorization {
      * @notice Disable recovery mode, which disables the special safe exit path for LPs.
      * @dev Protocol fees are not paid while in Recovery Mode, so it should only remain active for as long as strictly
      * necessary.
+     *
+     * This function will revert when called within a Vault context (i.e. in the middle of a join or an exit).
+     *
+     * Disabling recovery mode in derived contracts often triggers code that depends on the invariant or total supply,
+     * which may be calculated incorrectly in the middle of a join or an exit, because the state of the pool could be
+     * out of sync with the state of the Vault. The call to `ensureNotInVaultContext` in `VaultReentrancyLib` ensures
+     * this is safe to call.
+     *
+     * See https://forum.balancer.fi/t/reentrancy-vulnerability-scope-expanded/4345 for reference.
      */
     function disableRecoveryMode() external override authenticate {
         // Some derived contracts respond to disabling recovery mode with state changes (e.g., related to protocol fees,
         // or otherwise ensuring that enabling and disabling recovery mode has no ill effects on LPs). When called
         // outside of recovery mode, these state changes might lead to unexpected behavior.
         _ensureInRecoveryMode();
 
+        VaultReentrancyLib.ensureNotInVaultContext(_getVault());
```

Comment:
This should be accompanied by a test.

Reply:
Side comment (maybe for a follow-up PR): we are doing this here and in `ProtocolFeeCache` (which is derived from `RecoveryMode`) without a modifier.

Wouldn't it be more consistent to define the modifier here, and use it in both contracts? That way the code would be the same both in pools and in these auxiliary contracts.